File: 678-valid-parenthesis-string/valid-parenthesis-string.js
/**
 * @param {string} s
 * @return {boolean}
 */
var checkValidString = function(s) {
        let min = 0;
        let max =0;
        for (let i =0; i< s.length;i++){
            if(s[i] == '('){
                min+=1;
                max+=1;
            }else if (s[i]== ')'){
                min-=1;
                max-=1;
            }else{
                min-=1;
                max+=1;
            }
            if(min < 0){
                min = 0 ; 
            }
            if(max < 0 ){
                return false;
            }
        }
        return min == 0 ;
};

File: 678-valid-parenthesis-string/valid-parenthesis-string.java
class Solution {
    public boolean checkValidString(String s) {
        int min = 0;
        int max = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                min += 1;
                max += 1;
            } else if (s.charAt(i) == ')') {
                min -= 1;
                max -= 1;
            } else { // s.charAt(i) == '*'
                min -= 1;
                max += 1;
            }

            // Ensure min doesn't go negative
            if (min < 0) {
                min = 0;
            }

            // If max is negative, too many ')' so it's invalid
            if (max < 0) {
                return false;
            }
        }

        // If min is 0, all parentheses are balanced
        return min == 0;
    }
}
